File: src/tools/grep.ts
import { execFile } from 'child_process';
import { promisify } from 'util';
import { z } from 'zod';
import { tool } from './define.js';

const execFileP = promisify(execFile);

export const grepTool = tool({
  name: 'grep',
  description: 'Search file contents by regex. Uses ripgrep when available. Returns up to 100 matches.',
  inputSchema: z.object({
    pattern: z.string().describe('Regex pattern to search for'),
    path: z.string().optional().describe('Directory or file to search (default: cwd)'),
    glob: z.string().optional().describe('File filter, e.g. "*.ts"'),
    ignoreCase: z.boolean().optional(),
  }),
  execute: async ({ pattern, path, glob, ignoreCase }) => {
    try {
      const args = ['--line-number', '--no-heading', '--max-count=100', '--color=never'];
      if (ignoreCase) args.push('--ignore-case');
      if (glob) args.push('--glob', glob);
      args.push(pattern, path ?? '.');
      const { stdout } = await execFileP('rg', args, { maxBuffer: 1024 * 1024 });
      const matches = stdout.split('\n').filter(Boolean).slice(0, 100).map((line) => {
        const m = line.match(/^([^:]+):(\d+):(.*)$/);
        return m ? { file: m[1], line: Number(m[2]), content: m[3] } : { content: line };
      });
      return { matches, count: matches.length };
    } catch (err) {
      const e = err as { code?: number; stdout?: string; message: string };
      if (e.code === 1) return { matches: [], count: 0 };
      if (e.code === 'ENOENT' as unknown as number) {
        return { error: 'ripgrep (rg) not found. Install with `brew install ripgrep`.' };
      }
      return { error: e.message };
    }
  },
});

File: src/tools/grep.rs
use async_trait::async_trait;
use serde::Deserialize;
use serde_json::{json, Value};

use super::{PathPolicy, Tool};

pub struct GrepTool {
    pub path_policy: PathPolicy,
}

#[derive(Deserialize)]
struct Args {
    pattern: String,
    #[serde(default)]
    path: Option<String>,
    #[serde(default)]
    glob: Option<String>,
    #[serde(default, rename = "ignoreCase")]
    ignore_case: Option<bool>,
}

#[async_trait]
impl Tool for GrepTool {
    fn name(&self) -> &'static str {
        "grep"
    }
    fn description(&self) -> &'static str {
        "Search file contents by regex. Uses ripgrep when available. Returns up to 100 matches."
    }
    fn input_schema(&self) -> Value {
        json!({
            "type": "object",
            "properties": {
                "pattern": { "type": "string", "description": "Regex pattern to search for" },
                "path": { "type": "string", "description": "Directory or file to search (default: cwd)" },
                "glob": { "type": "string", "description": "File filter, e.g. \"*.ts\"" },
                "ignoreCase": { "type": "boolean" }
            },
            "required": ["pattern"]
        })
    }
    fn require_approval(&self, args: &Value) -> bool {
        args.get("path")
            .and_then(Value::as_str)
            .map(|p| self.path_policy.require_prompt_for_path(p))
            .unwrap_or(false)
    }
    async fn execute(&self, args: Value) -> Value {
        let args: Args = match serde_json::from_value(args) {
            Ok(a) => a,
            Err(e) => return json!({ "error": format!("invalid args: {e}") }),
        };
        let mut cmd = tokio::process::Command::new("rg");
        cmd.args([
            "--line-number",
            "--no-heading",
            "--max-count=100",
            "--color=never",
        ]);
        if args.ignore_case.unwrap_or(false) {
            cmd.arg("--ignore-case");
        }
        if let Some(g) = &args.glob {
            cmd.args(["--glob", g]);
        }
        cmd.arg(&args.pattern);
        let requested = args.path.unwrap_or_else(|| ".".into());
        if let Some(error) = self.path_policy.deny_path(&requested) {
            return json!({ "error": error });
        }
        let resolved = self.path_policy.resolve(&requested);
        cmd.arg(resolved.normalized);
        let output = match cmd.output().await {
            Ok(o) => o,
            Err(e) if e.kind() == std::io::ErrorKind::NotFound => {
                return json!({ "error": "ripgrep (rg) not found. Install with `brew install ripgrep`." });
            }
            Err(e) => return json!({ "error": e.to_string() }),
        };
        let code = output.status.code().unwrap_or(-1);
        if code == 1 {
            return json!({ "matches": [], "count": 0 });
        }
        if code != 0 {
            let stderr = String::from_utf8_lossy(&output.stderr);
            return json!({ "error": stderr.trim() });
        }
        let stdout = String::from_utf8_lossy(&output.stdout);
        let matches: Vec<Value> = stdout
            .split('\n')
            .filter(|l| !l.is_empty())
            .take(100)
            .map(|line| {
                let parts: Vec<&str> = line.splitn(3, ':').collect();
                if parts.len() == 3 {
                    if let Ok(n) = parts[1].parse::<u32>() {
                        return json!({
                            "file": parts[0],
                            "line": n,
                            "content": parts[2],
                        });
                    }
                }
                json!({ "content": line })
            })
            .collect();
        let count = matches.len();
        json!({ "matches": matches, "count": count })
    }
}
